core/src/main/java/io/grpc/internal/TransportTracer.java

```diff
@@ -0,0 +1,212 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import io.grpc.Status;
+import io.grpc.StreamTracer;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A class for gathering statistics about a transport. This is an experimental feature.
+ */
+public final class TransportTracer {
+  // streamsStarted happens serially, so a volatile is sufficient
+  private volatile long streamsStarted;
+  private long lastStreamCreatedTimeMsec;
+  private final AtomicLong streamsSucceeded = new AtomicLong();
+  private final AtomicLong streamsFailed = new AtomicLong();
+  private final AtomicLong messagesSent = new AtomicLong();
+  private final AtomicLong messagesReceived = new AtomicLong();
+  private final AtomicLong keepAlivesSent = new AtomicLong();
+  private final AtomicLong lastMessageSentTimeMsec = new AtomicLong();
+  private final AtomicLong lastMessageReceivedTimeMsec = new AtomicLong();
+  private volatile Callable<Integer> localFlowControlPollable;
+  private volatile Callable<Integer> remoteFlowControlPollable;
+
+  private final StreamTracer streamTracer = new StreamTracer() {
+    @Override
+    public void streamClosed(Status status) {
+      if (status.isOk()) {
+        streamsSucceeded.incrementAndGet();
+      } else {
+        streamsFailed.incrementAndGet();
+      }
+    }
+
+    @Override
+    public void outboundMessage(int seqNo) {
+      messagesSent.incrementAndGet();
+      updateMsecTimestamp(lastMessageSentTimeMsec);
+    }
+
+    @Override
+    public void inboundMessage(int seqNo) {
+      messagesReceived.incrementAndGet();
+      updateMsecTimestamp(lastMessageReceivedTimeMsec);
+    }
+  };
+
+  /**
+   * Returns a {@link StreamTracer} that can be installed on each stream created on the transport.
+   * The stats of each stream will be aggregated together on this TransportTracer.
+   */
+  public StreamTracer getStreamTracer() {
+    return streamTracer;
+  }
+
+  /**
+   * Called by the transport to report a stream has started. For clients, this happens when a header
+   * is sent. For servers, this happens when a header is received. This method must be called from
+   * only one thread, but the resulting stats may be read from any thread.
+   */
+  public void reportStreamStarted() {
+    // @SuppressWarnings is not allowed on an increment statement, using another var here.
+    long newStreamsStarted = streamsStarted + 1;
+    streamsStarted = newStreamsStarted;
```

Comment:
You might consider using the atomic increment function here.  The read and the write count as two barriers, where the atomic increment counts as just one.  My gut says atomic would be about the same speed.

Reply:
Added an unsychronized counter whose value we copy to the volatile, resulting in 1 volatile write.